File: sort.py
"""
NAME 
    sort - sorting and related algorithms 

DESCRIPTION 
    This module implements a few sorting and related algorithms. 

    * shuffle  - shuffle an array in-place
    * insort   - insertion sort 
    * qksort   - quick sort
    * mgsort   - merge sort 
    * qkselect - quick select
    * part3    - Dijkstra's 3-way partitioning

FUNCTIONS
    shuffle(nums)
        Shuffle an array in-place via Knuth's shuffling.

    insort(nums)
        Sort an array in-place into ascending order via insertion sort.

    qksort(nums)
        Sort an array in-place into ascending order via quick sort.

    mgsort(nums)
        Sort an array in-place into ascending order via merge sort.

    qkselect(nums, k)
        Return kth smallest element via quick select. 

    part3(nums)
        Partition an array into 3 parts via Dijkstra's 3-way partitioning.
"""


def shuffle(nums: List[int]) -> None: 
    """Shuffle an array in-place via Knuth's shuffling."""
    for i in range(1, len(nums)): 
        ii = randint(0, i)
        if ii != i: nums[ii], nums[i] = nums[i], nums[ii]


def insort(nums: List[int]) -> None:
    """Sort an array in-place into ascending order via insertion sort."""
    for i in range(1, len(nums)):
        for j in range(i, 0, -1):
            if nums[j-1] <= nums[j]: break
            nums[j-1], nums[j] = nums[j], nums[j-1]


def qksort(nums: List[int]) -> None:
    """Sort an array in-place into ascending order via quick sort."""
    shuffle(nums)                
        
    def sort(lo, hi): 
        """Sort nums[lo:hi] via quick sort."""
        if lo + 1 == hi: return 
        i, j = lo+1, hi-1
        while i <= j: 
            if nums[i] < nums[lo]: i += 1
            elif nums[j] > nums[lo]: j -= 1
            else: 
                nums[i], nums[j] = nums[j], nums[i]
                i += 1
                j -= 1
        nums[lo], nums[j] = nums[j], nums[lo]
        sort(lo, j)
        sort(j+1, hi)
        
    sort(0, len(nums))


def mgsort(nums: List[int]) -> None:
    """Sort an array in-place into ascending order via merge sort."""
    
    def sort(nums, aux, lo, hi): 
        """Sort nums[lo:hi] via merge sort."""
        if lo+1 >= hi: return 
        mid = lo + hi >> 1
        sort(aux, nums, lo, mid)
        sort(aux, nums, mid, hi)
        i, j = lo, mid
        for k in range(lo, hi): 
            if j >= hi or i < mid and aux[i] < aux[j]: 
                nums[k] = aux[i]
                i += 1
            else: 
                nums[k] = aux[j]
                j += 1
    
    sort(nums, nums.copy(), 0, len(nums))


def qkselect(nums: List[int], k: int) -> int:
    """Return kth smallest element via quick select."""
    shuffle(nums)

    def part(lo, hi):
        """Return partition of nums[lo:hi]."""
        i, j = lo+1, hi-1
        while i <= j: 
            if nums[i] < nums[lo]: i += 1
            elif nums[lo] < nums[j]: j -= 1
            else: 
                nums[i], nums[j] = nums[j], nums[i]
                i += 1
                j -= 1
        nums[lo], nums[j] = nums[j], nums[lo]
        return j 

    lo, hi = 0, len(nums)
    while lo < hi: 
        mid = part(nums, lo, hi)
        if mid < k-1: lo = mid + 1
        elif mid == k-1: return nums[mid]
        else: hi = mid


def part3(nums: List[int]) ->None: 
    """Partition an array into 3 parts via Dijkstra's 3-way partitioning."""
    lo, mid, hi = 0, 0, len(nums)-1
    while mid <= hi: 
        if nums[mid] == 0: 
            nums[lo], nums[mid] = nums[mid], nums[lo]
            lo += 1
            mid += 1
        elif nums[mid] == 1: mid += 1
        else: 
            nums[hi], nums[mid] = nums[mid], nums[hi]
            hi -= 1

File: javascript/sort.js
/*
NAME
    sort - sorting and related algorithms

DESCRIPTION
    This module implements a few sorting and related algorithms.

    * insort    insertion sort
    * mgsort    merge sort
    * part      partition
    * part3     Dijkstra's 3-way partitioning
    * qkselect  quick select
    * qksort    quick sort
    * shuffle   shuffle an array in-place

FUNCTIONS

    insort(nums)
        Sort an array in-place into ascending order via insertion sort.

    mgsort(nums)
        Sort an array in-place into ascending order via merge sort.

    part(nums, lo, hi)
        Partition subarray of nums from lo (inclusive) to hi (exclusive).

    part3(nums)
        Partition an array into 3 parts via Dijkstra's 3-way partitioning.

    qkselect(nums, k)
        Return kth smallest element via quick select.

    qksort(nums)
        Sort an array in-place into ascending order via quick sort.

    shuffle(nums)
        Shuffle an array in-place via Knuth's shuffling.
*/

function insort(nums) {
    /*insertion sort*/
    const n = nums.length;
    for (let i = 1; i < n; ++i)
        for (let j = i; j > 0; --j) {
            if (nums[j-1] <= nums[j]) break;
            [nums[j-1], nums[j]] = [nums[j], nums[j-1]];
        }
}


function mgsort(nums) {
    /*merge sort*/

    function sort(nums, aux, lo, hi) {
        if (lo+1 < hi) {
            const mid = Math.floor((lo+hi)/2);
            sort(aux, nums, lo, mid);
            sort(aux, nums, mid, hi);
            let i = lo, j = mid;
            for (let k = lo; k < hi; ++k)
                if (j >= hi || i < mid && aux[i] < aux[j])
                    nums[k] = aux[i++];
                else
                    nums[k] = aux[j++];
        }
    }

    sort(nums, nums.slice(), 0, nums.length);
}


function part(nums, lo, hi) {
    let i = lo+1, j = hi-1;
    while (i <= j) {
        if (nums[i] < nums[lo]) ++i;
        else if (nums[j] > nums[lo]) --j;
        else [nums[i++], nums[j--]] = [nums[j], nums[i]];
    }
    [nums[lo], nums[j]] = [nums[j], nums[lo]];
    return j;
}


function part3(nums) {
    /*Dijkstra's 3-way partition*/
    const n = nums.length;
    let lo = 0, mid = 0, hi = n-1;
    while (mid <= hi)
        if (nums[mid] == 0)
            [nums[lo++], nums[mid++]] = [nums[mid], nums[lo]];
        else if (nums[mid] == 1) ++mid;
        else [nums[hi--], nums[mid]] = [nums[mid], nums[hi]];
}


function qkselect(nums, k) {
    /*quick select*/
    shuffle(nums);
    let lo = 0, hi = nums.length;
    while (lo < hi) {
        const mid = part(nums, lo, hi);
        if (mid < k-1) lo = mid+1;
        else if (mid == k-1) return nums[mid];
        else hi = mid;
    }
}


function qksort(nums, lo=0, hi=-1) {
    /*quick sort by Tony Hoare*/
    if (hi == -1) hi = nums.length;
    if (lo+1 < hi) {
        const mid = part(nums, lo, hi);
        qksort(nums, lo, mid);
        qksort(nums, mid+1, hi);
    }
}


function shuffle(nums) {
    /*Knuth's shuffling*/
    const n = nums.length;
    for (let i = 1; i < n; ++i) {
        const ii = Math.floor(Math.random() * (i+1));
        [nums[ii], nums[i]] = [nums[i], nums[ii]];
    }
}
